### PR TITLE
Closing opened connections

### LINKED ISSUE #96: SQLDisconnect never called
Hi,

I'm using odbc_fdw to connect to a Teradata database.
Each time I run a query, three connections to the Teradata database are opened and the connections remain open until the session ends.
So if I run 3 queries in the same session:

```
fdw=> select * from lvl_teradata.table1 limit 1;
 STAB | PRATICA | FIDO | SUB | PER_AMM | TIP_RAT | NUM_RAT 
------+---------+------+-----+---------+---------+---------
(0 rows)

fdw=> select * from lvl_teradata.table1 limit 1;
 STAB | PRATICA | FIDO | SUB | PER_AMM | TIP_RAT | NUM_RAT 
------+---------+------+-----+---------+---------+---------
(0 rows)

fdw=> select * from lvl_teradata.table1 limit 1;
 STAB | PRATICA | FIDO | SUB | PER_AMM | TIP_RAT | NUM_RAT 
------+---------+------+-----+---------+---------+---------
(0 rows)
```

9 connections to Teradata will be open:

```
tcp        0      0 10.129.218.138:45278    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:36370    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:39168    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:36876    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:36382    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:56412    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:36888    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:46476    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
tcp        0      0 10.129.218.138:39156    10.242.31.130:1025      ESTABLISHED 44557/postgres: fdw 
```

Debugging odbc_fdw I see that odbc_connection, in my case, is called first by **odbcGetForeignRelSize**, then by **odbcEstimateCosts** and finally by **odbcBeginForeignScan**.

**odbcGetForeignRelSize** and **odbcEstimateCosts** call **odbcGetTableSize** in turn.
At the end of the **odbcGetTableSize** function there are the following lines of code:

https://github.com/CartoDB/odbc_fdw/blob/49579acb28793482fd1b7e54e1addd066be18686/odbc_fdw.c#L878-L895

**SQLDisconnect** will never be called because the dbc pointer is freed by **SQLFreeHandle (SQL_HANDLE_DBC, dbc)** and **dbc = NULL**.
I think SQLDisconnect needs to be moved before **SQLFreeHandle (SQL_HANDLE_DBC, dbc)**.

Furthermore, there is no closure after the end of the table scan (**odbcEndForeignScan**?).

Thanks,
Matteo

### PR DESCRIPTION
**Pull request by manmorjim**
This fixes #96 by closing opened connections when it is getting the size
of the table/query by executing COUNT and when it finishes the foreign
scan iteration.

### CODE REVIEW

**Comment by Algunenano** [comment]
2 things:
* Can we review the rest of the calls to `odbc_connection` and make sure those connections are closed too?
```
  826:  odbc_connection(options, &env, &dbc);
1051:  odbc_connection(&options, &env, &dbc);
1351:  odbc_connection(&options, &env, &dbc);
1949:  odbc_connection(&options, &env, &dbc);
2006:  odbc_connection(&options, &env, &dbc);
2104:  odbc_connection(&options, &env, &dbc);
```
* Do we need to disconnect as part of the error handling, i.e before calling `ereport` or `elog(ERROR`)?

**Comment by manmorjim** [comment]
> * Do we need to disconnect as part of the error handling, i.e before calling `ereport` or `elog(ERROR`)?

:thinking: I think so. As part of the revision I'll check also if all the connections are closed when some error occurred.

File: odbc_fdw.c [hunk 1]
```diff
@@ -415,6 +417,27 @@ odbc_connection(odbcFdwOptions* options, SQLHENV *env, SQLHDBC *dbc)
...
+	elog_debug("connection opened >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
+}
+
+/*
+ * Close the ODBC connection
+ */
+static void
+odbc_disconnection(SQLHENV *env, SQLHDBC *dbc)
+{
+	if (dbc && *dbc)
```
Comment (Algunenano) [inline odbc_fdw.c]: This is suspicious and likely what's causing crashes on Windows builds. You are setting the contents of `dbc` to NULL, but is `SQLHDBC` declared as a NULLable type in the ODBC standard (I'm guessing it comes from there)? 

In any case, the fact that we are asking ourselves this question means that this approach is wrong. Either always use a pointer or ensure you only call the disconnect once so you don't need the NULL guards.

**Comment by Algunenano** [comment]
I think you need to check the output of both the connect and disconnect calls. It might be that we are getting errors on them.

File: odbc_fdw.c [hunk 2]
```diff
@@ -1116,8 +1116,8 @@ Datum odbc_tables_list(PG_FUNCTION_ARGS)
 		datafctx->currentRow = currentRow;
 		SRF_RETURN_NEXT(funcctx, result);
 	} else {
-		odbc_disconnection(&env, &dbc);
 		SRF_RETURN_DONE(funcctx);
+		odbc_disconnection(&env, &dbc);
```
Comment (Algunenano) [inline odbc_fdw.c]: This will never be called since `SRF_RETURN_DONE` will return from the function.

Comment (manmorjim) [reply odbc_fdw.c]: Fixed. This function uses SRF and I was trying to disconnect it using wrong handlers. I stored these handlers into the `user_fctx` and using them to disconnect.

File: odbc_fdw.c [hunk 3]
```diff
@@ -414,6 +417,30 @@ odbc_connection(odbcFdwOptions* options, SQLHENV *env, SQLHDBC *dbc)
 	ret = SQLDriverConnect(*dbc, NULL, (SQLCHAR *) conn_str.data, SQL_NTS,
 	                       OutConnStr, 1024, &OutConnStrLen, SQL_DRIVER_COMPLETE);
 	check_return(ret, "Connecting to driver", dbc, SQL_HANDLE_DBC);
+	elog_debug("connection opened >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
```
Comment (Algunenano) [inline odbc_fdw.c]: I think this can be cleaned up (we can keep the message, but the `>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>` should go away).

Comment (Algunenano) [reply odbc_fdw.c]: Same for closing the connection

Comment (manmorjim) [reply odbc_fdw.c]: done!

**Comment by rafatower** [comment]
Jenkins tests are failing apparently due to mysql installation and should be easy to fix. Other than that, it'd be cool to merge and deploy this.

**Comment by jgoizueta** [comment]
# Broken tests

Tests spontaneously broke in Travis because of an update of the mysql 5.7 packages (we are installing mysql-server-core-5.7 mysql-client-5.7 mysql-server-5.7).

It the previous working tests the base image packages needn't an update:

```
mysql-client-5.7 is already the newest version (5.7.28-0ubuntu0.16.04.2).
mysql-server-5.7 is already the newest version (5.7.28-0ubuntu0.16.04.2).
mysql-server-core-5.7 is already the newest version (5.7.28-0ubuntu0.16.04.2).
```

But now they're updated

```
The following packages will be upgraded:
  mysql-client-5.7 mysql-server-5.7 mysql-server-core-5.7
```

And the update was failing:

```
Setting up mysql-client-5.7 (5.7.29-0ubuntu0.16.04.1) ...
Setting up mysql-server-core-5.7 (5.7.29-0ubuntu0.16.04.1) ...
Setting up mysql-server-5.7 (5.7.29-0ubuntu0.16.04.1) ...
insserv: warning: current start runlevel(s) (empty) of script `mysql' overrides LSB defaults (2 3 4 5).
insserv: warning: current stop runlevel(s) (0 1 2 3 4 5 6) of script `mysql' overrides LSB defaults (0 1 6).
mysql_upgrade: Got error: 2002: Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2) while connecting to the MySQL server
Upgrade process encountered error and will not continue.
mysql_upgrade failed with exit status 11
dpkg: error processing package mysql-server-5.7 (--configure):
 subprocess installed post-installation script returned error exit status 1
```

This is fixed by efd8ba1  and a0a8a43 by installing the pages *after* starting the server.

But then, access to MySQL fails with 

```
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
```

Which is fixed by 44b530f by changing the `root` authentication 😅

**Comment by jgoizueta** [comment]
@Algunenano could you give your blessing to this? I'll proceed to release and deploy this if you agree.

**Comment by Algunenano** [comment]
LGTM. I don't think we've ever updated odbc_fdw without also upgrading PG, so good luck with that.

**Comment by rafatower** [comment]
@jgoizueta is this deployed?

**Comment by jgoizueta** [comment]
This is released in v0.5.1 but not deployed, AFAIK we still have v0.4.0 in production

**Pull request merged**